backend/test_observer/controllers/test_results/filter_test_results.py

```diff
@@ -0,0 +1,186 @@
+    ).having(func.count(TestExecutionMetadata.value) == 0)
+
+    # Step 4: Exclude test executions with unmatched categories
+    # Exclude any test execution that has at least one
+    # unmatched category.
+    exclusion_condition = ~exists(
+        select(1)
+        .select_from(unmatched_categories.subquery())
+        .where(unmatched_categories.c.id == TestExecution.id)
+    )
+
+    return exclusion_condition
+
+
+def build_query_filters_and_joins(
+    filters: TestResultSearchFilters,
+) -> tuple[list[ColumnElement[bool]], set[str]]:
+    query_filters: list[ColumnElement[bool]] = []
+    joins_needed = set()
+
+    joins_needed.add("test_execution")
+
+    if filters.families is not None:
+        query_filters.append(Artefact.family.in_(filters.families))
+        joins_needed.update(["test_execution", "artefact_build", "artefact"])
+
+    if filters.environments is not None:
+        query_filters.append(Environment.name.in_(filters.environments))
+        joins_needed.update(["test_execution", "environment"])
+
+    if filters.test_cases is not None:
+        query_filters.append(TestCase.name.in_(filters.test_cases))
+        joins_needed.add("test_case")
+
+    if filters.template_ids is not None:
+        query_filters.append(TestCase.template_id.in_(filters.template_ids))
+        joins_needed.add("test_case")
+
+    if filters.execution_metadata is not None:
+        query_filters.append(filter_execution_metadata(filters.execution_metadata))
+        joins_needed.add("execution_metadata")
+
+    if filters.issues is not None:
+        query_filters.append(
+            TestResult.id.in_(
+                select(IssueTestResultAttachment.test_result_id)
+                .join(IssueTestResultAttachment.issue)
+                .where(Issue.id.in_(filters.issues))
+            )
+        )
+
+    if filters.from_date is not None:
+        query_filters.append(TestExecution.created_at >= filters.from_date)
+        joins_needed.add("test_execution")
+
+    if filters.until_date is not None:
+        query_filters.append(TestExecution.created_at <= filters.until_date)
+        joins_needed.add("test_execution")
+
+    return query_filters, joins_needed
+
+
+def apply_joins(query: Select, joins_needed: set[str]) -> Select:
+    if "test_execution" in joins_needed:
+        query = query.join(TestResult.test_execution)
+
+    if "environment" in joins_needed:
+        query = query.join(TestExecution.environment)
+
+    if "artefact_build" in joins_needed:
+        query = query.join(TestExecution.artefact_build)
+
+    if "artefact" in joins_needed:
+        query = query.join(ArtefactBuild.artefact)
+
+    if "test_case" in joins_needed:
+        query = query.join(TestResult.test_case)
+
+    if "execution_metadata" in joins_needed:
+        query = query.join(TestExecution.execution_metadata)
+
+    return query
+
+
+def filter_test_results(query: Select, filters: TestResultSearchFilters) -> Select:
```

Comment:
This is purely a refactor/moving things around so it can be reused by the bulk attach endpoint. Functionality for the `GET /test-results` endpoint should be exactly the same.